Use the Copy CompositeOperator when copying canvas into its layer.

mattwoodrow opened the pull request:
#### 876cd0f5d2135fb3b17077390274658f1b586c1c
<pre>
Use the Copy CompositeOperator when copying canvas into its layer.
<a href="https://bugs.webkit.org/show_bug.cgi?id=253614">https://bugs.webkit.org/show_bug.cgi?id=253614</a>

Reviewed by Simon Fraser.

When the canvas buffer is the only draw into the layer, we should be able to just directly copy the pixels instead of blending.
This also means we can skip clearing the back buffer.

* Source/WebCore/html/HTMLCanvasElement.cpp:
(WebCore::HTMLCanvasElement::paint):
* Source/WebCore/html/HTMLCanvasElement.h:
* Source/WebCore/platform/graphics/GraphicsLayer.cpp:
(WebCore::GraphicsLayer::GraphicsLayer):
* Source/WebCore/platform/graphics/GraphicsLayer.h:
(WebCore::GraphicsLayer::paintUsingCompositeCopy const):
(WebCore::GraphicsLayer::setPaintUsingCompositeCopy):
(WebCore::GraphicsLayer::canCopy const): Deleted.
(WebCore::GraphicsLayer::setCanCopy): Deleted.
* Source/WebCore/platform/graphics/ca/GraphicsLayerCA.h:
* Source/WebCore/platform/graphics/ca/PlatformCALayerClient.h:
(WebCore::PlatformCALayerClient::platformCALayerPaintUsingCompositeCopy const):
(WebCore::PlatformCALayerClient::platformCALayerCanCopy const): Deleted.
* Source/WebCore/rendering/RenderHTMLCanvas.cpp:
(WebCore::RenderHTMLCanvas::paintReplaced):
* Source/WebCore/rendering/RenderLayer.cpp:
(WebCore::RenderLayer::RenderLayer):
* Source/WebCore/rendering/RenderLayer.h:
(WebCore::RenderLayer::paintUsingCompositeCopy const):
(WebCore::RenderLayer::setPaintUsingCompositeCopy):
(WebCore::RenderLayer::canCopy const): Deleted.
(WebCore::RenderLayer::setCanCopy): Deleted.
* Source/WebCore/rendering/RenderLayerBacking.cpp:
(WebCore::RenderLayerBacking::updateConfiguration):
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerBackingStore.mm:
(WebKit::RemoteLayerBackingStore::drawInContext):

Canonical link: <a href="https://commits.webkit.org/261484@main">https://commits.webkit.org/261484@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/7f0d65c2527f378aa7f3cf13a168c470fa6e8e57

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/111813 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/20941 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/90/builds/427 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/87/builds/3582 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/120515 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/115884 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/22297 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/12009 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/86/builds/3329 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/117578 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/16562 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/99707 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/104827 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/98514 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/88/builds/283 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/35/builds/45533 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/13401 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/89/builds/284 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/92456 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/13905 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/84/builds/9712 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/19342 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/52286 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/7992 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/15883 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

smfr commented on Source/WebCore/platform/graphics/GraphicsLayer.h:
```diff
@@ -670,6 +670,9 @@ class GraphicsLayer : public RefCounted<GraphicsLayer> {
     virtual bool isGraphicsLayerTextureMapper() const { return false; }
     virtual bool isCoordinatedGraphicsLayer() const { return false; }
 
+    bool paintUsingCompositeCopy() const { return m_paintUsingCompositeCopy; }
```
`canPaintUsingCompositeCopy()` or `shouldPaintUsingCompositeCopy()` (this function isn't doing a paint).

smfr commented on Source/WebCore/platform/graphics/ca/GraphicsLayerCA.h:
```diff
@@ -237,6 +237,8 @@ class GraphicsLayerCA : public GraphicsLayer, public PlatformCALayerClient {
     WEBCORE_EXPORT bool platformCALayerUseCSS3DTransformInteroperability() const override;
     WEBCORE_EXPORT void platformCALayerLogFilledVisibleFreshTile(unsigned) override;
     bool platformCALayerContainsBitmapOnly(const PlatformCALayer*) const override { return client().layerContainsBitmapOnly(this); }
+    bool platformCALayerPaintUsingCompositeCopy() const override { return paintUsingCompositeCopy(); }
```
can/shouldPaint

smfr commented on Source/WebCore/platform/graphics/ca/PlatformCALayerClient.h:
```diff
@@ -73,6 +73,8 @@ class PlatformCALayerClient {
 
     virtual bool platformCALayerContainsBitmapOnly(const PlatformCALayer*) const { return false; }
 
+    virtual bool platformCALayerPaintUsingCompositeCopy() const { return false; }
```
can/should paint

smfr commented on Source/WebCore/rendering/RenderLayer.h:
```diff
@@ -785,6 +785,8 @@ class RenderLayer : public CanMakeWeakPtr<RenderLayer> {
 
     bool isolatesBlending() const { return hasNotIsolatedBlendingDescendants() && isCSSStackingContext(); }
     
+    bool paintUsingCompositeCopy() const { return m_paintUsingCompositeCopy; }
```
can/should

smfr commented on Source/WebCore/rendering/RenderLayer.h:
```diff
@@ -1144,6 +1146,8 @@ class RenderLayer : public CanMakeWeakPtr<RenderLayer> {
     
     bool computeHasVisibleContent() const;
 
+    void setPaintUsingCompositeCopy(bool copy) { m_paintUsingCompositeCopy = copy; }
```
setCan/Should

mattwoodrow committed:
Use the Copy CompositeOperator when copying canvas into its layer.
https://bugs.webkit.org/show_bug.cgi?id=253614

Reviewed by Simon Fraser.

When the canvas buffer is the only draw into the layer, we should be able to just directly copy the pixels instead of blending.
This also means we can skip clearing the back buffer.

* Source/WebCore/html/HTMLCanvasElement.cpp:
(WebCore::HTMLCanvasElement::paint):
* Source/WebCore/html/HTMLCanvasElement.h:
* Source/WebCore/platform/graphics/GraphicsLayer.cpp:
(WebCore::GraphicsLayer::GraphicsLayer):
* Source/WebCore/platform/graphics/GraphicsLayer.h:
(WebCore::GraphicsLayer::paintUsingCompositeCopy const):
(WebCore::GraphicsLayer::setPaintUsingCompositeCopy):
(WebCore::GraphicsLayer::canCopy const): Deleted.
(WebCore::GraphicsLayer::setCanCopy): Deleted.
* Source/WebCore/platform/graphics/ca/GraphicsLayerCA.h:
* Source/WebCore/platform/graphics/ca/PlatformCALayerClient.h:
(WebCore::PlatformCALayerClient::platformCALayerPaintUsingCompositeCopy const):
(WebCore::PlatformCALayerClient::platformCALayerCanCopy const): Deleted.
* Source/WebCore/rendering/RenderHTMLCanvas.cpp:
(WebCore::RenderHTMLCanvas::paintReplaced):
* Source/WebCore/rendering/RenderLayer.cpp:
(WebCore::RenderLayer::RenderLayer):
* Source/WebCore/rendering/RenderLayer.h:
(WebCore::RenderLayer::paintUsingCompositeCopy const):
(WebCore::RenderLayer::setPaintUsingCompositeCopy):
(WebCore::RenderLayer::canCopy const): Deleted.
(WebCore::RenderLayer::setCanCopy): Deleted.
* Source/WebCore/rendering/RenderLayerBacking.cpp:
(WebCore::RenderLayerBacking::updateConfiguration):
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerBackingStore.mm:
(WebKit::RemoteLayerBackingStore::drawInContext):

Canonical link: https://commits.webkit.org/261484@main
